v0/destinations/user/utils.js

```diff
@@ -0,0 +1,670 @@
+ * Remaining keys : keys which is not included in webapp configuration mapping and not included in source-dest keys file
+ * @param {*} traits
+ * @param {*} sourceKeys
+ * @returns
+ */
+const getRemainingAttributes = (traits, sourceKeys) => {
+  const properties = {};
+  const keys = Object.keys(traits);
+  keys.forEach(key => {
+    if (!sourceKeys.includes(key)) {
+      properties[key] = traits[key];
+    }
+  });
+  return properties;
+};
+
+/**
+ * Returns the customAttributes (user, company and event custom attributes) + remaining attributes
+ * Remaining attributes : keys which is not included in webapp configuration mapping and not included in source-dest keys file
+ * @param {*} attributesMap
+ * @param {*} properties
+ * @param {*} excludeKeys
+ * @returns
+ */
+const getAttributes = (attributesMap, properties, excludeKeys) => {
+  const sourceKeys = excludeKeys;
+  const data = {};
+  const attributesMapKeys = Object.keys(attributesMap);
+
+  attributesMapKeys.forEach(key => {
+    if (properties[key]) {
+      const destinationAttributeName = attributesMap[key];
+      data[destinationAttributeName] = properties[key];
+      sourceKeys.push(key);
+    }
+  });
+
+  const remainingAttributes = getRemainingAttributes(properties, sourceKeys);
+  return { ...data, ...remainingAttributes };
+};
+
+/**
+ * Returns the identify call payload
+ * @param {*} destination
+ * @param {*} commonUserPropertiesPayload
+ * @param {*} message
+ * @returns
+ */
+const prepareIdentifyPayload = (
+  destination,
+  commonUserPropertiesPayload,
+  message
+) => {
+  const payload = commonUserPropertiesPayload;
+  const { userAttributesMapping } = destination.Config;
+  const userAttributesMap = getHashFromArray(
+    userAttributesMapping,
+    "from",
+    "to",
+    false
+  );
+  const traits = getIdentifyTraits(message);
+  const customUserAttributes = getAttributes(
+    userAttributesMap,
+    traits,
+    identifySourceKeys
+  );
+  return {
+    ...payload,
+    ...customUserAttributes
+  };
+};
+
+/**
+ * validating the group call payload
+ * if any of the required parameter is not present then we will throw an error
+ * @param {*} destination
+ * @param {*} message
+ * @param {*} commonCompanyProperties
+ * @returns
+ */
+const prepareCreateOrUpdateCompanyPayload = (
+  destination,
+  message,
+  commonCompanyProperties
+) => {
+  let commonCompanyPropertiesPayload = commonCompanyProperties;
+  const { companyAttributesMapping } = destination.Config;
+  const companyAttributesMap = getHashFromArray(
+    companyAttributesMapping,
+    "from",
+    "to",
+    false
+  );
+  const traits = getFieldValueFromMessage(message, "traits");
+  const customCompanyAttributes = getAttributes(
+    companyAttributesMap,
+    traits,
+    groupSourceKeys
+  );
+  commonCompanyPropertiesPayload = {
+    ...commonCompanyPropertiesPayload,
+    ...customCompanyAttributes
+  };
+  if (commonCompanyPropertiesPayload.address) {
```

Comment:
let us also check if it is of type object. because address can come as just a single string as well. In that case object keys will not work.